### PR TITLE
[feature#3356] alert-spi support DingTalk&WeChat

### LINKED ISSUE #3356: [Feature][SPI] Implement DingTalk Alert Plugin
Implement DingTalk Alert Plugin

#3049 

I will finish it.

### PR DESCRIPTION
**Pull request by CalvinKirs**
this closes #3356

TBD:need test,delete old class

### CODE REVIEW

**Review by gaojun2048 — CHANGES_REQUESTED** [review]
unit test failed

**Comment by CalvinKirs** [comment]
TBD:weChat

**Comment by CalvinKirs** [comment]
Whether to enable other alert methods, we need to further discuss, we may need to specify a specification, and judge on the alarm server

**Comment by CalvinKirs** [comment]
> Whether to enable other alert methods, we need to further discuss, we may need to specify a specification, and judge on the alarm server

For example, plug-in name + enable? The historical version of a WeChat is determined by the configuration parameter enterprise.wechat.enabl

**Review by gaojun2048 — COMMENTED** [review]
> > Whether to enable other alert methods, we need to further discuss, we may need to specify a specification, and judge on the alarm server
> 
> For example, plug-in name + enable? The historical version of a WeChat is determined by the configuration parameter enterprise.wechat.enabl

In the feature, If you want to send WeChat, you need select WeChat plugin and input the params to create a WeChat plugin instance, then add the plugin instance to a alert group , then select this alert group at workflow define.
When the workflow failed the AlertServer will find the alert group this workflow selected and then get all the plugin instance in this alert group and send message by it.

**Comment by CalvinKirs** [comment]
TBD，There are some details that are still being revised. When the revision is completed, I will review it myself, and then update the status of the PR, and then you can review it

File: dolphinscheduler-alert-plugin/dolphinscheduler-alert-dingtalk/src/main/java/org/apache/dolphinscheduler/plugin/alert/dingtalk/DingTalkAlertChannel.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            paramsMap.put(param.getName(), param.getValue().toString());
+        }
+        AlertResult alertResult = new AlertResult();
+        alertResult.setStatus(Boolean.toString(Boolean.TRUE));
+        DingTalkSender dingTalkSender = new DingTalkSender(paramsMap);
+        try {
+            dingTalkSender.sendDingTalkMsg(alertData.getTitle(), alertData.getContent());
+        } catch (IOException e) {
+            alertResult.setStatus(Boolean.toString(Boolean.FALSE));
+            logger.error(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (CalvinKirs) [reply dolphinscheduler-alert-plugin/dolphinscheduler-alert-dingtalk/src/main/java/org/apache/dolphinscheduler/plugin/alert/dingtalk/DingTalkAlertChannel.java:58]: Thank you for your suggestion. I will judge the response parameters of DingTalk.

**Comment by CalvinKirs** [comment]
I added specific WeChat configurations to the unit test of the WeChat alert plug-in, and you can directly test the specific effects. Before the PR is merged, I will delete these configurations

File: dolphinscheduler-alert-plugin/dolphinscheduler-alert-dingtalk/src/main/java/org/apache/dolphinscheduler/plugin/alert/dingtalk/DingTalkSender.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            this.errmsg = errmsg;
+        }
+    }
+
+    private static AlertResult checkSendDingTalkSendMsgResult(String result) {
+        AlertResult alertResult = new AlertResult();
+        alertResult.setStatus("false");
+
+        if (null == result) {
+            alertResult.setMessage("we chat alert send error");
```
Comment (gaojun2048) [inline dolphinscheduler-alert-plugin/dolphinscheduler-alert-dingtalk/src/main/java/org/apache/dolphinscheduler/plugin/alert/dingtalk/DingTalkSender.java:184]: wechat alert ?

File: dolphinscheduler-alert-plugin/dolphinscheduler-alert-dingtalk/src/main/java/org/apache/dolphinscheduler/plugin/alert/dingtalk/DingTalkSender.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            logger.info("send ding talk msg error,ding talk server resp is null");
+            return alertResult;
+        }
+        DingTalkSendMsgResponse sendMsgResponse = JSONUtils.parseObject(result, DingTalkSendMsgResponse.class);
+        if (null == sendMsgResponse) {
+            alertResult.setMessage("we chat send fail");
```
Comment (gaojun2048) [inline dolphinscheduler-alert-plugin/dolphinscheduler-alert-dingtalk/src/main/java/org/apache/dolphinscheduler/plugin/alert/dingtalk/DingTalkSender.java:190]: we chat ?

File: dolphinscheduler-alert-plugin/dolphinscheduler-alert-wechat/src/main/java/org/apache/dolphinscheduler/plugin/alert/wechat/WeChatAlertConstants.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ */
+public class WeChatAlertConstants {
+
+    static final String MARKDOWN_QUOTE = ">";
+
+    static final String MARKDOWN_ENTER = "\n";
+
+    static final String CHARSET = "UTF-8";
+
+    static final String WE_CHAT_PUSH_URL = "https://qyapi.weixin.qq.com/cgi-bin/message/send?access_token={token}";
```
Comment (gaojun2048) [inline dolphinscheduler-alert-plugin/dolphinscheduler-alert-wechat/src/main/java/org/apache/dolphinscheduler/plugin/alert/wechat/WeChatAlertConstants.java:31]: can we add qyapi.weixin.qq.com in apache project?

File: pom.xml [hunk 5]
```diff
@@ -950,6 +950,11 @@
                         <include>**/plugin/alert/email/ExcelUtilsTest.java</include>
                         <include>**/plugin/alert/email/MailUtilsTest.java</include>
                         <include>**/plugin/alert/email/template/DefaultHTMLTemplateTest.java</include>
+                        <include>**/plugin/alert/dingtalk/DingTalkSenderTest.java</include>
+                        <include>**/plugin/alert/dingtalk/DingTalkAlertChannelFactoryTest.java</include>
+                        <include>**/plugin/alert/dingtalk/DingTalkAlertChannelFactoryTest.java</include>
```
Comment (gaojun2048) [inline pom.xml:955]: DingTalkAlertChannelFactoryTest.java is repeated.

**Review by gaojun2048 — APPROVED** [review]
Please take a look at the comment I added.
Thank you!

File: dolphinscheduler-alert-plugin/dolphinscheduler-alert-wechat/src/main/java/org/apache/dolphinscheduler/plugin/alert/wechat/WeChatAlertConstants.java (hunk 4 above)
Comment (CalvinKirs) [reply dolphinscheduler-alert-plugin/dolphinscheduler-alert-wechat/src/main/java/org/apache/dolphinscheduler/plugin/alert/wechat/WeChatAlertConstants.java:31]: Of course, I have confirmed this problem

**Review by gaojun2048 — APPROVED** [review]
+1

**Pull request merged**